core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -0,0 +1,214 @@
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.distribution.DistributionInfo;
+import org.infinispan.distribution.DistributionManager;
+import org.infinispan.factories.KnownComponentNames;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.metadata.impl.IracMetadata;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.impl.VoidResponseCollector;
+import org.infinispan.util.ExponentialBackOff;
+import org.infinispan.util.concurrent.AggregateCompletionStage;
+import org.infinispan.util.concurrent.CompletableFutures;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.infinispan.xsite.XSiteBackup;
+import org.infinispan.xsite.irac.DefaultIracManager;
+import org.infinispan.xsite.irac.IracExecutor;
+
+@Scope(Scopes.NAMED_CACHE)
+public class DefaultIracTombstoneManager implements IracTombstoneManager {
+
+   @Inject DistributionManager distributionManager;
+   @Inject RpcManager rpcManager;
+   @Inject CommandsFactory commandsFactory;
+   private final Map<Object, TombstoneData> tombstoneMap;
+   private final IracExecutor iracExecutor;
+   private final Collection<XSiteBackup> asyncBackups;
+
+   public DefaultIracTombstoneManager(Configuration configuration) {
+      this.iracExecutor = new IracExecutor(this::performCleanup);
+      this.asyncBackups = DefaultIracManager.asyncBackups(configuration);
+      this.tombstoneMap = new ConcurrentHashMap<>();
+   }
+
+   @Inject
+   public void inject(@ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService executorService,
+                      @ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor blockingExecutor) {
+      // using the inject method here in order to decrease the class size
+      iracExecutor.setBackOff(ExponentialBackOff.NO_OP);
+      iracExecutor.setExecutor(blockingExecutor);
+      // TODO configure?
+      executorService.scheduleAtFixedRate(this::startCleanupTombstone, 30, 30, TimeUnit.SECONDS);
+   }
+
+   public void storeTombstone(int segment, Object key, IracMetadata metadata) {
+      tombstoneMap.put(key, new TombstoneData(segment, metadata));
+   }
+
+   @Override
+   public void storeTombstoneIfAbsent(int segment, Object key, IracMetadata metadata) {
+      if (metadata == null) {
+         return;
+      }
+      tombstoneMap.putIfAbsent(key, new TombstoneData(segment, metadata));
+   }
+
+   @Override
+   public IracMetadata getTombstone(Object key) {
+      TombstoneData data = tombstoneMap.get(key);
+      return data == null ? null : data.getMetadata();
+   }
+
+   @Override
+   public void removeTombstone(Object key, IracMetadata iracMetadata) {
+      if (iracMetadata == null) {
+         return;
+      }
+      remove(key, new TombstoneData(-1, iracMetadata));
```

Comment:
Should this use the segment for the key?

Reply:
the segment is useful for `performCleanup()`, which is executed for every key every 30 seconds :) 
In this case, the removal is confirmed to happen (if the tombstone does not change) and I don't need the segment.

Reply:
Hmmm, do you expect keys to stay in the tombstones map for a lot more than 30 seconds?

Reply:
it can happen if the connection between sites is broken.
The tombstone has to stay until the "take-offline" configuration kicks in.

Reply:
I wouldn't expect the take-offline time-to-wait to be much longer than 30s either, and IMO it's more important to save a few bytes by not allocating a wrapper object (assuming the IracMetadata object is also small).